plugin/pkg/auth/authenticator/password/keystone/keystone.go

```diff
@@ -45,18 +50,74 @@ func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticatePassword(usernam
 		glog.Info("Failed: Starting openstack authenticate client")
 		return nil, false, errors.New("Failed to authenticate")
 	}
-
+//FIXME user_id, not user_name...
 	return &user.DefaultInfo{Name: username}, true, nil
 }
 
-// NewKeystoneAuthenticator returns a password authenticator that validates credentials using openstack keystone
-func NewKeystoneAuthenticator(authURL string) (*KeystoneAuthenticator, error) {
-	if !strings.HasPrefix(authURL, "https") {
-		return nil, errors.New("Auth URL should be secure and start with https")
+func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticateToken(value string) (user.Info, bool, error) {
+//FIXME kfox1111: To enhance performance, cache admin token/client.
+	cfg, provider, err := loadConfig(keystoneAuthenticator.configPath)
+	if err != nil {
+		return nil, false, errors.New("Internal error getting the Keystone provider")
+	}
+
+	client, err := openstack.NewIdentityAdminV3(provider, gophercloud.EndpointOpts{
+		Region: cfg.Global.Region,
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	response, err := client.Request("GET", client.ServiceURL("auth", "tokens"), gophercloud.RequestOpts{
+	        MoreHeaders: map[string]string{"X-Subject-Token": value},
+	        OkCodes:     []int{200},
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	defer response.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		glog.V(3).Infof("Cannot get HTTP response body from keystone token validate: %v", err)
+		return nil, false, err
+	}
+	obj := struct{ Token struct { User struct { Id string }}}{}
+	err = json.Unmarshal(bodyBytes, &obj)
+	if err != nil {
+		return nil, false, err
+	}
+	return &user.DefaultInfo{Name: obj.Token.User.Id}, response.StatusCode == 200, nil
```

Comment:
We need to think about it now, before we commit to supporting the feature. If not, its a recipe for a lot of pain in the future. Keystone didn't put in enough forethought into it in the past, and that's why they are desperately trying to get rid of v2 now. Because its just not flexible enough. Its taken them many years to do so.

Almost every new OpenStack cloud these days is multidomain. They use them for service accounts, trusts, etc. One domain thats sql backed for service accounts, one domain for ldap, one domain for heat trusts, one for sahara trusts, magnum trusts, etc.

So it is highly likely that a user may want to do all of the following on a Magnum launched k8s:
 * interact with k8s via the kubectl command using their horizon provided openstackrc file for credentials
 * interact with k8s via heat (which uses a trust user in a different domain which the user delegates access to his/her behalf)
 * interact with k8s via the horizon plugin

All of these may involve different domains, and all have tokens in common in their implementations.

On my cloud, once k2k federation is complete, our setup will most likely be the following:
* One keystone with an kerberos domain as a k2k identity provider
* Two separate regions, each with their own Keystone with sql backed service users and heat domain users. Setup as a k2k service provider. This allows keystone load to be sharded nicely.

The user would login through Horizon via the identity provider, federate to one of the regions, launch a kubernetes cluster via magnum and the cluster would point back to the region's keystone. k2k federation this way is entirely transparent to the user.

Then, to use the cli:
The user would download the openrc file which would include:
domain,project,identity_keystone_url, region name,auth_type_krb5 (note, no username/password)

The openstack client for authentication will:
contact the identity provider keystone, authenticate with it using the ticket granting ticket from the user's krb5 login session to get the saml assertion, and the region specified keystone.
pass the saml assertion to the region keystone and get a token back valid for that region.

The token can then be passed to kubectl to to send to the api-server to send back to the region keystone for validation.

This kind of setup won't work well with username/password at all. But the authentication is seamless for users. The krb5 tgt is fetched automatically when they login to their own machine.

We don't want to support all the ways Keystone works inside the api-server. The keystone token mechanism provides an abstraction so that Keystone can be modified as needed to scale, add features, enhance security, etc without having to change the api all the time.

Doing Keystone username/password inside the server forces a very narrow surface of Keystone that is going to be so painfully restrictive, so I think its better to just skip entirely.

You can leave it as an alternative to Tokens, but I think its going to get a lot of issues raised against it in the same vain as the "keystone v3 isn't supported" issue. "kerberos isn't supported", "k2k isn't supported", etc.

There will need to be a solution for the k8s dashboard, sure, but I don't think username/password on the server is the way to solve that. k8s dashboard is going to need to be aware of many more authentication types Keystone does in the wild or else k8s dashboard won't be able to be widely used in a lot of the enterprise environments OpenStack's supporting. Authentication in k8s dashboard itself will need to be plugable.